File: web/lib/resampling.ts
import { OHLCData } from "@/actions/data-actions"

// Helper to parse timeframe string into seconds
// e.g., "1m" -> 60, "1h" -> 3600, "240" -> 14400
export function parseTimeframeToSeconds(tf: string): number {
    // Check for resolution string (number only)
    if (/^\d+$/.test(tf)) {
        return parseInt(tf, 10) * 60 // Default to minutes
    }

    // Strict case matching for units: m (min), h (hour), D (day), W (week), M (month), Y (year)
    // We allow 'd' and 'w' to leniently map to Day/Week, but 'm' MUST be minute and 'M' MUST be Month
    const match = tf.match(/^(\d+)(m|h|d|w|M|y)$/i)
    if (!match) return 0

    const [, numStr, unit] = match
    const num = parseInt(numStr, 10)

    // Use raw unit for m/M distinction
    switch (unit) {
        case 'm': return num * 60
        case 'h':
        case 'H': return num * 60 * 60
        case 'd':
        case 'D': return num * 60 * 60 * 24
        case 'w':
        case 'W': return num * 60 * 60 * 24 * 7
        case 'M': return num * 60 * 60 * 24 * 30 // Approximate month
        case 'y':
        case 'Y': return num * 60 * 60 * 24 * 365 // Approximate year
        default: return 0
    }
}

// Main resampling function
// Aggregates lower timeframe data into higher timeframe buckets
export function resampleOHLC(data: OHLCData[], fromTF: string, toTF: string): OHLCData[] {
    const fromSeconds = parseTimeframeToSeconds(fromTF)
    const toSeconds = parseTimeframeToSeconds(toTF)

    // Validation
    if (fromSeconds === 0 || toSeconds === 0) {
        console.error(`Invalid timeframes: ${fromTF} -> ${toTF}`)
        return []
    }

    if (toSeconds <= fromSeconds) {
        // Target TF is smaller or equal, return original data
        // (Downsampling not supported, equal doesn't need resampling)
        return data
    }

    // Daily/Weekly handling: STRICTLY prohibit resampling to D/W/M
    // User requested to rely on native files for these to ensure settlement time accuracy
    // Check for D, W, M in the suffix
    if (toTF.match(/[DWM]$/) || toTF.match(/[dw]$/)) {
        console.warn(`Resampling to ${toTF} is not supported (Daily/Weekly require native data)`)
        return []
    }

    if (data.length === 0) return []

    const resampled: OHLCData[] = []
    let currentBucket: OHLCData | null = null
    let bucketEndTime = Number.NaN // Initialize to NaN so first comparison (valid_ts !== NaN) is always true

    for (const candle of data) {
        // Calculate which bucket this candle belongs to
        // We align to the start of the bucket (e.g., 10:00:00 for 10:00-10:05 bucket)
        const timestamp = candle.time
        const bucketStart = Math.floor(timestamp / toSeconds) * toSeconds

        // If this is a new bucket
        if (bucketStart !== bucketEndTime) {
            // Push previous bucket if complete
            if (currentBucket) {
                resampled.push(currentBucket)
            }

            // Start new bucket
            currentBucket = {
                time: bucketStart,
                open: candle.open,
                high: candle.high,
                low: candle.low,
                close: candle.close,
                volume: candle.volume || 0
            }
            bucketEndTime = bucketStart
        } else if (currentBucket) {
            // Aggregate into current bucket
            currentBucket.high = Math.max(currentBucket.high, candle.high)
            currentBucket.low = Math.min(currentBucket.low, candle.low)
            currentBucket.close = candle.close // Close is always the last candle's close
            currentBucket.volume = (currentBucket.volume || 0) + (candle.volume || 0)
        }
    }

    // Push final bucket
    if (currentBucket) {
        resampled.push(currentBucket);
        // If weekly aggregation, add empty bucket for next week start
        if (isWeekly) {
            const nextWeekStart = currentBucket.time + 7 * 86400;
            resampled.push({ time: nextWeekStart, open: 0, high: 0, low: 0, close: 0, volume: 0 });
        }
    }

    return resampled;
}

// Function to check if we can resample from source to target
export function canResample(fromTF: string, toTF: string): boolean {
    const fromSeconds = parseTimeframeToSeconds(fromTF)
    const toSeconds = parseTimeframeToSeconds(toTF)

    // Valid if target is larger
    if (toSeconds <= fromSeconds) return false

    // AND target is not Daily/Weekly/Monthly (D, W, M)
    // We strictly check for 'D', 'W', 'M', 'd', 'w' but ALLOW 'm' (minutes) AND number-only strings
    if (toTF.match(/[DWM]$/)) return false // Uppercase D, W, M forbidden
    if (toTF.match(/[dw]$/)) return false  // Lowercase d, w forbidden (just in case)

    // 'm' (minutes) and raw resolution strings ("240") are allowed!
    return true
}

// Helper to resample Daily/Weekly/Monthly/Yearly on the client for fallback purposes
export function resampleDataForWMY(data: OHLCData[], targetTimeframe: string): OHLCData[] {
    if (data.length === 0) return data

    // Parse timeframe
    const isWeekly = targetTimeframe.toUpperCase().endsWith('W')
    const isMonthly = targetTimeframe.toUpperCase().endsWith('M')
    const isYearly = targetTimeframe.toUpperCase().endsWith('Y')

    let months = 1
    if (isMonthly) {
        const match = targetTimeframe.match(/^(\d+)M$/i)
        if (match) {
            months = parseInt(match[1], 10)
        }
    }
    let years = 1
    if (isYearly) {
        const match = targetTimeframe.match(/^(\d+)Y$/i)
        if (match) {
            years = parseInt(match[1], 10)
        }
    }

    const resampled: OHLCData[] = []
    let currentBucket: OHLCData | null = null
    let bucketEndTime = Number.NaN

    for (const candle of data) {
        const date = new Date(candle.time * 1000)
        let bucketStart = 0

        if (isWeekly) {
            const weekSeconds = 7 * 86400;
            bucketStart = Math.floor(candle.time / weekSeconds) * weekSeconds;
        } else if (isMonthly) {
            const year = date.getUTCFullYear()
            const month = date.getUTCMonth()
            const bucketMonth = Math.floor(month / months) * months
            const firstOfMonth = new Date(Date.UTC(year, bucketMonth, 1, 0, 0, 0, 0))
            bucketStart = Math.floor(firstOfMonth.getTime() / 1000)
        } else if (isYearly) {
            const year = date.getUTCFullYear()
            const bucketYear = Math.floor(year / years) * years
            const firstOfYear = new Date(Date.UTC(bucketYear, 0, 1, 0, 0, 0, 0))
            bucketStart = Math.floor(firstOfYear.getTime() / 1000)
        } else {
            // Fallback to simple seconds-based bucketing if it's something else
            const toSeconds = parseTimeframeToSeconds(targetTimeframe)
            if (toSeconds <= 0) return data
            bucketStart = Math.floor(candle.time / toSeconds) * toSeconds
        }

        if (bucketStart !== bucketEndTime) {
            if (currentBucket) resampled.push(currentBucket)
            currentBucket = { ...candle, time: bucketStart }
            bucketEndTime = bucketStart
        } else if (currentBucket) {
            currentBucket.high = Math.max(currentBucket.high, candle.high)
            currentBucket.low = Math.min(currentBucket.low, candle.low)
            currentBucket.close = candle.close
            currentBucket.volume = (currentBucket.volume || 0) + (candle.volume || 0)
        }
    }
    if (currentBucket) resampled.push(currentBucket)
    return resampled
}

File: api/features/shared/resampling.py
"""
OHLC Resampling Service

This module ports the TypeScript resampling logic from web/lib/resampling.ts
to Python to ensure consistent aggregation between frontend chart display
and backend backtesting.

IMPORTANT: This logic MUST match the TypeScript implementation exactly!
Any changes here should be mirrored in resampling.ts and vice versa.
"""

import re
import pandas as pd
from typing import Optional


def parse_timeframe_to_seconds(tf: str) -> int:
    """
    Parse timeframe string into seconds.
    
    Examples:
        "1m" -> 60
        "5m" -> 300
        "1h" -> 3600
        "1D" -> 86400
        "240" -> 14400 (numeric = minutes)
    
    Must match TypeScript: parseTimeframeToSeconds()
    """
    # Check for resolution string (number only) - default to minutes
    if re.match(r'^\d+$', tf):
        return int(tf) * 60
    
    # Match: number + unit (m, h, d, w, D, W, M)
    match = re.match(r'^(\d+)([mhdwMDW])$', tf, re.IGNORECASE)
    if not match:
        return 0
    
    num = int(match.group(1))
    unit = match.group(2)
    
    # Use exact case for m/M distinction
    if unit == 'm':
        return num * 60
    elif unit in ('h', 'H'):
        return num * 60 * 60
    elif unit in ('d', 'D'):
        return num * 60 * 60 * 24
    elif unit in ('w', 'W'):
        return num * 60 * 60 * 24 * 7
    elif unit == 'M':
        return num * 60 * 60 * 24 * 30  # Approximate month
    else:
        return 0


def can_resample(from_tf: str, to_tf: str) -> bool:
    """
    Check if resampling from source to target is allowed.
    
    Rules:
    - Target must be larger than source
    - Daily/Weekly/Monthly (D, W, M) targets are NOT allowed
      (must use native data files for accurate settlement times)
    
    Must match TypeScript: canResample()
    """
    from_seconds = parse_timeframe_to_seconds(from_tf)
    to_seconds = parse_timeframe_to_seconds(to_tf)
    
    if to_seconds <= from_seconds:
        return False
    
    # Forbid D, W, M targets
    if re.search(r'[DWM]$', to_tf):
        return False
    if re.search(r'[dw]$', to_tf):
        return False
    
    return True


def resample_ohlc(df: pd.DataFrame, from_tf: str, to_tf: str) -> pd.DataFrame:
    """
    Aggregate lower timeframe data into higher timeframe buckets.
    
    Aggregation rules (MUST match TypeScript exactly):
    - open: first candle's open
    - high: max of all highs
    - low: min of all lows
    - close: last candle's close
    - volume: sum of all volumes
    
    Must match TypeScript: resampleOHLC()
    
    Args:
        df: DataFrame with columns [time, open, high, low, close, volume]
        from_tf: Source timeframe (e.g., "1m")
        to_tf: Target timeframe (e.g., "5m")
    
    Returns:
        Resampled DataFrame
    """
    from_seconds = parse_timeframe_to_seconds(from_tf)
    to_seconds = parse_timeframe_to_seconds(to_tf)
    
    # Validation
    if from_seconds == 0 or to_seconds == 0:
        print(f"Invalid timeframes: {from_tf} -> {to_tf}")
        return pd.DataFrame()
    
    if to_seconds <= from_seconds:
        # Target is smaller or equal, return original
        return df.copy()
    
    # Prohibit resampling to D/W/M
    if re.search(r'[DWMdw]$', to_tf):
        print(f"Resampling to {to_tf} is not supported (Daily/Weekly require native data)")
        return pd.DataFrame()
    
    if df.empty:
        return df.copy()
    
    # Calculate bucket for each row
    df = df.copy()
    df['bucket'] = (df['time'] // to_seconds) * to_seconds
    
    # Aggregate
    resampled = df.groupby('bucket').agg({
        'time': 'first',  # Will be replaced with bucket
        'open': 'first',
        'high': 'max',
        'low': 'min',
        'close': 'last',
        'volume': 'sum'
    }).reset_index(drop=True)
    
    # Set time to bucket start
    resampled['time'] = df.groupby('bucket')['bucket'].first().values
    
    return resampled
